### ISSUE TITLE
HTML Doesn't Consistently Highlight Script and Style Tags

### ISSUE THREAD

**Issue by BrandonQDixon** [PrismJS/prism#3202]
**Information:**
- Prism version: 1.25.0
- Plugins: none
- Environment: Browser & Angular

**Description**
When rendering HTML contents, using either tags "language-html" or "language-markup", the style tags and css tags are not consistently rendered.

**Example**
In the screenshot below, we can see some JavaScript is being rendered properly, but the CSS is not:
![image](https://user-images.githubusercontent.com/22106438/142915659-3d07fb27-6235-4a86-a6d0-9a5d9b6eeb14.png)

**Comment by RunDevelopment** [comment]
Hello @BrandonQDixon!

What does "not consistently rendered" mean? Please provide more information so we can understand and reproduce the problem. If possible, could you please provide a minimal example reproducing your issue?

**Comment by BrandonQDixon** [comment]
Hello, I'll try to setup a minimal example in a moment. In the meantime, I'll describe that the JavaScript would sometimes render and sometimes not, and the CSS never renders. The rest of the HTML document renders as normal.

**Comment by RunDevelopment** [comment]
Sounds like a race condition in the order languages are loaded. How do you [load Prism](https://prismjs.com/#basic-usage)? Did you download a bundle from [our download page](https://prismjs.com/download.html) and use that or do you load languages individually (e.g. using [Autoloader](https://prismjs.com/plugins/autoloader/))?

**Comment by BrandonQDixon** [comment]
I am wrapping this in my own Angular component and am loading scripts via the code below. It is loading min and core css, then the autoloader, then the line numbers; however: each promise.all block means each item within the group inside of the method call is loading at the same time. Also note: I've only seen this issue in this case, where the single code block actually has more than one language itself (html and css and javascript). I was using version 1.24.x before and updated in an effort to fix this, but that did not resolve.
` private async loadRes(): Promise<void> {
        if (BasicKitCodeBoxComponent.resLoaded) {
            return;
        }
        BasicKitCodeBoxComponent.resLoaded = true;
        await Promise.all([
            this.httpRequestService.loadCss("https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/themes/prism.min.css"),
            this.httpRequestService.loadScript("https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/components/prism-core.min.js")
        ]);
        await Promise.all([
            this.httpRequestService.loadScript("https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/plugins/autoloader/prism-autoloader.min.js"),
        ]);
        await Promise.all([
            this.httpRequestService.loadCss("https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/plugins/line-numbers/prism-line-numbers.min.css"),
            this.httpRequestService.loadScript("https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/plugins/line-numbers/prism-line-numbers.min.js")
        ]);
    }`

**Comment by RunDevelopment** [comment]
I see the problem now. Autoloader will only load the minimum number of languages necessary to highlight a piece of code. So if you give it some piece of markup code, it will only load the Markup language. It won't load CSS or JS.

The solution is to tell Autoloader to also load CSS and JS alongside Markup. This is done using [the `data-dependencies` attribute](https://prismjs.com/plugins/autoloader/).

The randomness around `<style>` tags being highlighted is most likely because you also have CSS code on the same page, right? Autoloader will load Markup and CSS in parallel. If Markup is loaded before CSS, then `<style>` tags won't be highlighted because Autoloader doesn't know that it was to wait for CSS as well since there is no `data-dependencies` attribute. If CSS was loaded before markup, then you got lucky and see highlighted `<style>` tags.

---

Also, a note on load performance: You can load Autoloader and Line numbers in parallel.

**Comment by BrandonQDixon** [comment]
Thank you for the information, you are correct in your assessment of the "randomness" and the solution in general. I've added "css,javascript" to data-dependencies and have also grouped the script loading and everything is now solved.

**Note to anyone who happens upon this:** I've seen references on other sites saying that html should be rendered with "language-markdown" class. This appears to be incorrect, as I could only arrive at this solution after switching back from that, to "language-html" as I've had it originally.

**Comment by RunDevelopment** [comment]
> everything is now solved

Glad I could help.

> html should be rendered with "language-markdown" class

Yeah, that's definitely not a solution but I understand why it might **appear** to work for some people. Markdown has a few dependencies and the Autoloader has to load those dependencies before Markdown. So Markdown takes longer to load than CSS/JS and therefore it is less likely to win the race against CSS/JS.

**Comment by BrandonQDixon** [comment]
Hello, unfortunately I've discovered some more intermittent behavior. I've published the site where we can observe the issue in action; the code box appears at the bottom: https://bytethisstore.com/articles/pg/simple-contact-form

I've also included the full component code for reference in case I've missed anything before.

```typescript
import { ChangeDetectionStrategy, ChangeDetectorRef, Component } from "@angular/core";
import { InjectableWindowModule, InjectableWindowService } from "@brandonqdixon/angular-frontend-common-injectable-window";
import { BaseCodeBoxComponent } from "../../abstract-components/code-box-component/code-box-component";
import { CodeBoxSelector } from "../../models/components/code-box-component";
import { HttpRequestService } from "@brandonqdixon/angular-frontend-common-http-request";
import { ClipboardService } from "@brandonqdixon/angular-frontend-common-clipboard";
import { BehaviorSubject, combineLatest } from "rxjs";
import { map } from "rxjs/operators";

@Component({
    selector: CodeBoxSelector,
    templateUrl: './code-box-component.html',
    styleUrls: ['./code-box-component.scss'],
    //changeDetection: ChangeDetectionStrategy.OnPush
    //encapsulation: ViewEncapsulation.ShadowDom
})
export class BasicKitCodeBoxComponent extends BaseCodeBoxComponent {

    static resLoaded = false;

    showButtonsRow$ = combineLatest([
        this.copyEnabled$,
        this.viewRawEnabled$
    ]).pipe(map(booleans => {
        return booleans.reduce((bool, boolean) => bool || boolean, false);
    }));

    COPY_MESSAGES = {
        copy: "Copy to Clipboard",
        copied: "Copied to Clipboard!"
    };
    copyMessage$ = new BehaviorSubject(this.COPY_MESSAGES.copy);

    displayCodes$ = new BehaviorSubject<any>(null);

    constructor(
        private injectableWindowService: InjectableWindowService,
        private httpRequestService: HttpRequestService,
        private clipboardService: ClipboardService
    ) {
        super();
    }

    async doNgOnInit() {
        await this.loadRes();

        this.registerSubscription(
            this.codes$.subscribe(codes => {
                this.displayCodes$.next(null);
                setTimeout(() => {
                    this.displayCodes$.next(codes);
                    const injectableWindow = this.injectableWindowService.window;
                    if ((injectableWindow as any).Prism) {
                        (injectableWindow as any).Prism.manual = true;
                        (injectableWindow as any).Prism.highlightAll();
                        setTimeout(() => {
                            (injectableWindow as any).Prism.highlightAll();
                        }, 200);
                        setTimeout(() => {
                            (injectableWindow as any).Prism.highlightAll();
                        }, 2000);
                    }
                }, 50)
            })
        );
    }

    copyToClipboard(str: string): void {
        this.copyMessage$.next(this.COPY_MESSAGES.copied);
        this.clipboardService.copyToClipboard(str);
        setTimeout(() => {
            this.copyMessage$.next(this.COPY_MESSAGES.copy);
        }, 1500);
    }

    viewRaw(str: string): void {
        const injWindow = this.injectableWindowService.window;
        const encodedStr = `data:text/plain;charset=utf-8,${encodeURIComponent(str)}`;
        const win = injWindow.open();
        win?.document.write(`<iframe src="${encodedStr}" frameborder="0" style="border:0; top:0px; left:0px; bottom:0px; right:0px; width:100%; height:100%;" allowfullscreen></iframe>`)
    }

    getLanguageCssClasses(language: string): string {
        let lang = "";
        switch (language.toLocaleLowerCase()) {
            case "c#":
                lang = "csharp";
                break;
            case "html":
                lang = "html";
                break;
            default:
                lang = language;
        }
        return `language-${lang}`
    }

    getDataDependencies(language: string): string {
        let deps: string[] = [];
        switch (language) {
            case "html":
                deps = deps.concat("css", "javascript");
                break;
        }
        return deps.join(",");
    }

    mapCodeLanguage(language: string): string {
        switch (language) {
            case "javascript":
                language = "JavaScript";
                break;
            case "typescript":
                language = "TypeScript";
                break;
            default:
                language = language.substring(0, 1).toUpperCase() + language.substring(1);
                break;
        }
        return language;
    }

    private async loadRes(): Promise<void> {
        if (BasicKitCodeBoxComponent.resLoaded) {
            return;
        }
        BasicKitCodeBoxComponent.resLoaded = true;
        await Promise.all([
            this.httpRequestService.loadCss("https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/themes/prism.min.css"),
            this.httpRequestService.loadScript("https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/components/prism-core.min.js")
        ]);
        await Promise.all([
            this.httpRequestService.loadScript("https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/plugins/autoloader/prism-autoloader.min.js"),
            this.httpRequestService.loadCss("https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/plugins/line-numbers/prism-line-numbers.min.css"),
            this.httpRequestService.loadScript("https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/plugins/line-numbers/prism-line-numbers.min.js")
        ]);
    }

}
```

```html
<ng-container *ngIf="displayCodes$ | async as codes">
    <div class="single-container" *ngIf="codes.length === 1; else MultipleCodes">
        <pre class="line-numbers">
            <code [ngClass]="getLanguageCssClasses(codes[0].language)" [attr.data-dependencies]="getDataDependencies(codes[0].language)">{{ codes[0].code }}</code>
        </pre>
        <div id="buttons-row" *ngIf="showButtonsRow$ | async">
            <action-button *ngIf="copyEnabled$ | async" (click)="copyToClipboard(codes[0].code)">{{ copyMessage$ | async
                }}</action-button>
            <action-button *ngIf="viewRawEnabled$ | async" (click)="viewRaw(codes[0].code)">View Raw</action-button>
        </div>
    </div>
    <ng-template #MultipleCodes>
        <input-tabbed-box-container>
            <input-tabbed-box *ngFor="let code of codes; let ind = index" [title]="mapCodeLanguage(code.language)">
                <pre
                    class="line-numbers"><code [ngClass]="getLanguageCssClasses(code.language)" [attr.data-dependencies]="getDataDependencies(code.language)">{{ code.code }}</code></pre>
                <div id="buttons-row" *ngIf="showButtonsRow$ | async">
                    <action-button *ngIf="copyEnabled$ | async" (click)="copyToClipboard(codes[ind].code)">{{
                        copyMessage$ | async }}
                    </action-button>
                    <action-button *ngIf="viewRawEnabled$ | async" (click)="viewRaw(codes[ind].code)">View Raw
                    </action-button>
                </div>
            </input-tabbed-box>
        </input-tabbed-box-container>
    </ng-template>
</ng-container>
```

**Comment by RunDevelopment** [comment]
Looking at your site, I can see the problem and it's Prism's fault. 

Autoloader does load HTML+JS+CSS but it loads them in the wrong order. This is because Prism has [3 types of dependencies](https://prismjs.com/extending.html#declaring-dependencies) but Autoloader only supports 1 (#1540). The other 2 kinds are somewhat rare (and optional), so it's typically not a huge issue. We likely won't fix this bug in v1 anymore because the planned dependency system in v2 will automatically fix this.

In the meantime, you can work around this issue by loading `prism/1.25.0/prism.min.js` instead of `prism/1.25.0/components/prism-core.min.js`. `prism.min.js` is a small bundle that includes Core, HTML, CSS, JS, C-like (required by JS), and the File Highlight plugin. Since all these languages are present in the bundle, Autoloader won't have to load them and you won't encounter this bug.

Sorry for the inconvenience. This bug slipped my mind.
